spark_log_parser/loaders/__init__.py

```diff
@@ -0,0 +1,282 @@
+            return True
+
+        filename = filename.lower()
+        return any([name in filename for name in FILE_SKIP_PATTERNS])
+
+    def read_tgz_archive(self, archive: tarfile.TarFile):
+        """
+
+        """
+        # TODO - validate size
+        size = 0
+        for tarinfo in archive:
+            if self.should_skip_file(tarinfo.name):
+                continue
+
+            if tarinfo.isdir():
+                continue
+
+            # TODO - Size assertions?
+            size += tarinfo.size
+
+            file_bytes = archive.extractfile(tarinfo)
+            yield from self.extract(Path(tarinfo.name), file_bytes)
+
+    def read_zip_archive(self, raw_archive_stream):
+        """
+
+        """
+
+        total_size = 0
+        for fname, fsize, chunks in stream_unzip(raw_archive_stream):
+            fname = fname.decode("utf-8")
+            should_skip = fsize == 0 or self.should_skip_file(fname)
+            if should_skip:
+                # We still need to exhaust the chunks for this part of the archive in order to
+                #  be able to consume the stream for the next part! If we don't do this, stream_unzip
+                #  will raise an UnfinishedIterationError, since we can't just skip ahead
+                for _ in chunks:
+                    continue
+                continue
+
+            # TODO - Size assertions?
+            if fsize is not None:
+                total_size += fsize
+
+            wrapped = ZipArchiveMemberWrapper(chunks)
+            yield from self.extract(Path(fname), wrapped)
+
+    @abc.abstractmethod
+    def read_file_stream(self, file: IOBase):
+        """
+        When we finally hit a file that no longer needs to be expanded/decompressed in some manner, this method
+        will be called to yield data from the file in some manner. This is here is order to allow concrete classes
+        to decide how they actually want to yield data from the file, whether that be raw chunks, as lines, or a blob
+        of the whole file (or something else entirely)
+        """
+
+    @abc.abstractmethod
+    def load_item(self, filepath: str | ParseResult):
+        """
+        Since the concrete Loader may be fetching data from disparate data sources, this method should be
+        responsible for grabbing the raw data stream for each file that may then be passed to extract().
+
+        Alternatively, if the file is present locally on the machine, just passing a filepath to extract()
+        will suffice
+        """
+
+    def extract_directory(self, directory: Path):
+        for path in directory.iterdir():
+            yield from self.extract(path)
+
+    def extract(self, filepath: Path, file_stream: BufferedIOBase = None):
+        """
+
+        """
+        if file_stream is None and filepath.is_dir():
+            yield from self.extract_directory(filepath)
+
+        to_open = file_stream if file_stream is not None else filepath
```

Comment:
Yeah, that's a good question. This was an attempt to just be able to pass the same variable to any of these `open` calls, but it really only works with `gzip`. I agree, this is just making it harder to follow - will remove

Reply:
Not that I'm defending this variable, but I do see a lot of places in the larger IO module that treat filepaths and `IOBase` instances as basically "identical", which is why I initially did this. For example, in `gzip.open` -

```
    The filename argument can be an actual filename (a str or bytes object), or
    an existing file object to read from or write to.
```

Reply:
Good point. It's nice that `gzip.open` behaves that way, and it feels idiomatic. If this method should only accept one or the other and not both, then I think it would be helpful to developers if that were evident in the signature.